File: src/socket/Buffer.cpp
//
// Created by ga78cat on 24.03.2021.
//

#include <comm/socket/Buffer.h>
#include <cassert>
#include <comm/socket/utils.h>
#include <cstring>
#include <iostream>

using namespace comm;
using namespace std;

Buffer::Buffer(unsigned long long int bufferSize) :
        buffer(nullptr), bufferSize(0), bufferContentSize(0), referenceBuffer(nullptr), useReferenceBuffer(false) {
    this->prepareBuffer(bufferSize);
}

Buffer::~Buffer() {
    this->reset();
}

void Buffer::reset() {
    delete[] this->buffer;
    this->buffer = nullptr;
    this->bufferSize = 0;
    this->bufferContentSize = 0;
    this->referenceBuffer = nullptr;
    this->useReferenceBuffer = false;
}

Buffer *Buffer::copy() const {
    auto *copy = new Buffer(this->bufferSize);

    assert (copy->bufferSize == this->bufferSize);
    memcpy(copy->buffer, this->buffer, this->bufferSize);

    copy->bufferContentSize = this->bufferContentSize;

    copy->referenceBuffer = this->referenceBuffer;
    copy->useReferenceBuffer = this->useReferenceBuffer;

    return copy;
}

void Buffer::setData(char *data, unsigned int dataSize, int start) {
    this->setData((const char *) data, dataSize, start);
}

void Buffer::setData(const char *data, unsigned int dataSize, int start) {
    this->setBufferContentSize(dataSize + start);
    memcpy(this->buffer + start, data, dataSize);
    this->useReferenceBuffer = false;
}

void Buffer::setReferenceToData(char *data, unsigned int dataSize) {
    this->referenceBuffer = data;
    this->bufferContentSize = dataSize;
    this->useReferenceBuffer = true;
}

void Buffer::setChar(char data, int position) {
    this->checkBufferContentSize(position + sizeof(char), true);
    this->buffer[position] = data;
    this->useReferenceBuffer = false;
}

void Buffer::setShort(short data, int position) {
    this->checkBufferContentSize(position + sizeof(short), true);
    shortToNetworkBytes(this->buffer, position, data);
    this->useReferenceBuffer = false;
}

void Buffer::setInt(int data, int position) {
    this->checkBufferContentSize(position + sizeof(int), true);
    intToNetworkBytes(this->buffer, position, data);
    this->useReferenceBuffer = false;
}

void Buffer::setLongLong(long long data, int position) {
    this->checkBufferContentSize(position + sizeof(long long), true);
    longLongToNetworkBytes(this->buffer, position, data);
    this->useReferenceBuffer = false;
}

void Buffer::setFloat(float data, int position) {
    this->checkBufferContentSize(position + sizeof(float), true);
    floatToNetworkBytes(this->buffer, position, data);
    this->useReferenceBuffer = false;
}

void Buffer::setDouble(double data, int position) {
    this->checkBufferContentSize(position + sizeof(double), true);
    doubleToNetworkBytes(this->buffer, position, data);
    this->useReferenceBuffer = false;
}

char Buffer::getChar(int position) {
    this->checkBufferContentSize(position + sizeof(char), false);
    return this->buffer[position];
}

short Buffer::getShort(int position) {
    this->checkBufferContentSize(position + sizeof(short), false);
    return networkBytesToShort(this->buffer, position);
}

int Buffer::getInt(int position) {
    this->checkBufferContentSize(position + sizeof(int), false);
    return networkBytesToInt(this->buffer, position);
}

long long Buffer::getLongLong(int position) {
    this->checkBufferContentSize(position + sizeof(long long), false);
    return networkBytesToLongLong(this->buffer, position);
}

float Buffer::getFloat(int position) {
    this->checkBufferContentSize(position + sizeof(float), false);
    return networkBytesToFloat(this->buffer, position);
}

double Buffer::getDouble(int position) {
    this->checkBufferContentSize(position + sizeof(double), false);
    return networkBytesToDouble(this->buffer, position);
}

bool Buffer::empty() const {
    return this->bufferContentSize == 0;
}

char *Buffer::getBuffer() {
    char *result = (this->useReferenceBuffer) ? this->referenceBuffer : this->buffer;
    this->useReferenceBuffer = false;
    return result;
}

unsigned long long int Buffer::getBufferSize() const {
    return this->bufferSize;
}

unsigned long long int Buffer::getBufferContentSize() const {
    return this->bufferContentSize;
}

void Buffer::setBufferContentSize(unsigned long long int _bufferContentSize) {
    this->prepareBuffer(_bufferContentSize);
    this->bufferContentSize = _bufferContentSize;
}

void Buffer::prepareBuffer(unsigned long long int desiredSize) {
    if (this->bufferSize < desiredSize) {
        char *oldBuffer = this->buffer;
        unsigned long long int oldSize = this->bufferSize;

        // cout << "Initialize new buffer!" << endl;
        this->buffer = new char[desiredSize];
        this->bufferSize = desiredSize;
        // cout << "Initialized new buffer (" << bufferLength << ")!" << endl;

        if (oldSize > 0) {
            memcpy(this->buffer, oldBuffer, oldSize * sizeof(*(this->buffer)));
        }
        memset(this->buffer + oldSize, 0, (this->bufferSize - oldSize) * sizeof(*(this->buffer)));

        delete[] oldBuffer;
    }
}

void Buffer::checkBufferContentSize(unsigned long long size, bool modifySize) {
    if (size > this->bufferContentSize) {
        if (modifySize) {
            this->prepareBuffer(size);
            this->bufferContentSize = size;
        } else {
            (*cerror) << "Size of " << size << " is greater than the current buffer content size " << this->bufferContentSize
                 << "! The next assertion will fail..." << endl;
        }
    }
    if (size > this->bufferContentSize) {
        cout << "Requested size = " << size << " vs. this->bufferContentSize = " << this->bufferContentSize << endl;
        assert (!modifySize);
    }
    assert (size <= this->bufferContentSize);
}

File: csharp/Comm/utils/Buffer.cs
using System;
using System.Diagnostics;

namespace Comm.utils {
    public class Buffer {
        public Buffer(ulong bufferSize = 0) {
            this.buffer = null;
            this.bufferSize = 0;
            this.bufferContentSize = 0;
            this.referenceBuffer = null;
            this.useReferenceBuffer = false;
            this.prepareBuffer(bufferSize);
        }

        ~Buffer() {
            this.reset();
        }

        public void reset() {
            this.buffer = null;
            this.bufferSize = 0;
            this.bufferContentSize = 0;
            this.referenceBuffer = null;
            this.useReferenceBuffer = false;
        }

        public Buffer copy() {
            Buffer copy = new Buffer(this.bufferSize);

            Debug.Assert(copy.bufferSize == this.bufferSize);
            Utils.memcpy(copy.buffer, 0, this.buffer, 0, this.bufferSize);

            copy.bufferContentSize = this.bufferContentSize;

            copy.referenceBuffer = this.referenceBuffer;
            copy.useReferenceBuffer = this.useReferenceBuffer;

            return copy;
        }

        public void setData(byte[] data, ulong dataSize, ulong start = 0, ulong dataStart = 0) {
            this.setBufferContentSize(dataSize + start);
            Utils.memcpy(this.buffer, start, data, dataStart, dataSize);
            this.useReferenceBuffer = false;
        }

        public void setReferenceToData(byte[] data, ulong dataSize) {
            this.referenceBuffer = data;
            this.bufferContentSize = dataSize;
            this.useReferenceBuffer = true;
        }

        public void setByte(byte data, int position) {
            this.buffer[position] = data;
        }

        public void setShort(short data, int position) {
            NetworkData.shortToNetworkBytes(this.buffer, position, data);
        }

        public void setInt(int data, int position) {
            NetworkData.intToNetworkBytes(this.buffer, position, data);
        }

        public void setLongLong(long data, int position) {
            NetworkData.longLongToNetworkBytes(this.buffer, position, data);
        }

        public void setFloat(float data, int position) {
            NetworkData.floatToNetworkBytes(this.buffer, position, data);
        }

        public void setDouble(double data, int position) {
            NetworkData.doubleToNetworkBytes(this.buffer, position, data);
        }

        public byte getByte(int position) {
            return buffer[position];
        }

        public short getShort(int position) {
            return NetworkData.networkBytesToShort(this.buffer, position);
        }

        public int getInt(int position) {
            return NetworkData.networkBytesToInt(this.buffer, position);
        }

        public long getLongLong(int position) {
            return NetworkData.networkBytesToLongLong(this.buffer, position);
        }

        public float getFloat(int position) {
            return NetworkData.networkBytesToFloat(this.buffer, position);
        }

        public double getDouble(int position) {
            return NetworkData.networkBytesToDouble(this.buffer, position);
        }

        public bool empty() {
            return (this.bufferContentSize == 0);
        }

        public byte[] getBuffer() {
            if (this.useReferenceBuffer) {
                this.useReferenceBuffer = false;
                return this.referenceBuffer;
            }
            return this.buffer;
        }

        public ulong getBufferSize() {
            return this.bufferSize;
        }

        public ulong getBufferContentSize() {
            return this.bufferContentSize;
        }

        public void setBufferContentSize(ulong _bufferContentSize) {
            this.prepareBuffer(_bufferContentSize);
            this.bufferContentSize = _bufferContentSize;
        }


        private void prepareBuffer(ulong desiredSize) {
            if (this.bufferSize < desiredSize) {
                byte[] oldBuffer = this.buffer;
                ulong oldSize = this.bufferSize;

                // cout << "Initialize new buffer!" << endl;
                this.buffer = new byte[desiredSize];
                this.bufferSize = desiredSize;
                // cout << "Initialized new buffer (" << bufferLength << ")!" << endl;

                Utils.memcpy(this.buffer, 0, oldBuffer, 0, oldSize * sizeof(byte));
                Utils.memset(this.buffer, oldSize, 0, (this.bufferSize - oldSize) * sizeof(byte));
            }
        }

        private void checkBufferContentSize(ulong size, bool modifySize) {
            if (size > this.bufferContentSize) {
                if (modifySize) {
                    this.prepareBuffer(size);
                    this.bufferContentSize = size;
                } else {
                    Console.WriteLine("Size of " + size + " is greater than the current buffer content size " + this.bufferContentSize + "! The next assertion will fail...\n");
                }
            }
            if (size > this.bufferContentSize) {
                Console.WriteLine("Requested size = " + size + " vs. this.bufferContentSize = " + this.bufferContentSize + "\n");
                Debug.Assert(!modifySize);
            }
            Debug.Assert(size <= this.bufferContentSize);
        }

        byte[] buffer, referenceBuffer;
        ulong bufferSize, bufferContentSize;
        bool useReferenceBuffer;
    };
}
